lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,239 @@
+            duration_s = 10,
+            period_ms = 16,
+            cpus = [cpu], # pin to big CPU, to focus on frequency selection
+            sched_policy = 'FIFO' # RT tasks have the boost holding feature so
+                                  # the frequency should be more stable, and we
+                                  # shouldn't go to max freq in Android
+        )
+        return rtapp_profile
+
+    def trace_window(self, trace):
+        """
+        Set the boundaries of the trace window to ``cpu_frequency`` events
+        before/after the task's start/end time
+        """
+        rta_start, rta_stop = super().trace_window(trace)
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        fdf = trace.df_events('cpu_frequency')
+        fdf = fdf[fdf.cpu == cpu]
+
+        # Find the frequency events before and after the task runs
+        freq_start = fdf[fdf.index < rta_start].index[-1]
+        freq_stop = fdf[fdf.index > rta_stop].index[0]
+
+        return (freq_start, freq_stop)
+
+    @requires_events('cpu_frequency')
+    def test_stune_frequency(self, freq_margin_pct=10) -> ResultBundle:
+        """
+        Test that frequency selection followed the boost
+
+        :param: freq_margin_pct: Allowed margin between estimated and measured
+            average frequencies
+        :type freq_margin_pct: int
+
+        Compute the expected frequency given the boost level and compare to the
+        real average frequency from the trace.
+        Check that the difference between expected and measured frequencies is
+        no larger than ``freq_margin_pct``.
+        """
+        kernel_version = self.plat_info['kernel']['version']
+        if kernel_version.parts[:2] < (4, 14):
+            self.get_logger().warning('This test requires the RT boost hold, but it may be disabled in {}'.format(kernel_version))
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        freqs = self.plat_info['freqs'][cpu]
+        max_freq = max(freqs)
+
+        # Estimate the target frequency, including sugov's margin, and round
+        # into a real OPP
+        boost = self.boost
+        target_freq = min(max_freq, max_freq * boost / 80)
+        target_freq = list(filter(lambda f: f >= target_freq, freqs))[0]
+
+	# Get the real average frequency
+        avg_freq = self.trace.analysis.frequency.get_average_cpu_frequency(cpu)
+
+        distance = abs(target_freq - avg_freq) * 100 / target_freq
+        res = ResultBundle.from_bool(distance < freq_margin_pct)
+        res.add_metric("target freq", target_freq, 'kHz')
+        res.add_metric("average freq", avg_freq, 'kHz')
+
+        return res
+
+class SchedTuneFrequencyTest(SchedTuneBase):
+    """
+    Runs multiple ``SchedTuneFreqItem`` tests at various boost levels ranging
+    from 20% to 100%, then checks all succedeed.
+    """
+    @classmethod
+    def _create_test_bundles(cls, target, res_dir, ftrace_coll):
+        for boost in range(20, 101, 20):
```

Comment:
is the sequence `20, 101, 20` intended, or should the last one be something else ?

Reply:
I want `20,40,60,80,100`. Perhaps I should list them explicitly rather than using `range()` here